File: 1137-n-th-tribonacci-number/1137-n-th-tribonacci-number.py
class Solution(object):
    def __init__(self):
        self.memo = {}
        
    def tribonacci(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n in self.memo:
            return self.memo[n]
        
        if n == 0:
            result = 0
        elif n == 1 or n == 2:
            result = 1
        else:
            result = self.tribonacci(n - 1) + self.tribonacci(n - 2) + self.tribonacci(n - 3)
        #memoize
        self.memo[n] = result
        return result

File: 1137-n-th-tribonacci-number/1137-n-th-tribonacci-number.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    int tribonacci(int n) {
        // Vector to store memoized Tribonacci values
        memo.resize(n + 1, -1);

        return tribonacciMemo(n);
    }

private:
    std::vector<int> memo;

    int tribonacciMemo(int n) {
        // Check if the result for 'n' is already memoized
        if (memo[n] != -1) {
            return memo[n];
        }

        // Base cases
        if (n == 0) {
            return memo[n] = 0;
        } else if (n == 1 || n == 2) {
            return memo[n] = 1;
        }

        // Recursive calculation
        return memo[n] = tribonacciMemo(n - 1) + tribonacciMemo(n - 2) + tribonacciMemo(n - 3);
    }
};
